cirq/linalg/transformations.py

```diff
@@ -304,3 +307,109 @@ def partial_trace(tensor: np.ndarray,
     left_indices = [keep_map[i] if i in keep_set else i for i in range(ndim)]
     right_indices = [ndim + i if i in keep_set else i for i in left_indices]
     return np.einsum(tensor, left_indices + right_indices)
+
+
+def wavefunction_partial_trace(wavefunction: np.ndarray,
+                               keep_indices: List[int],
+                               atol: Union[int, float] = 1e-8) -> Tuple:
+    """Computes a mixture representing the partial trace of a wavefunction.
+
+    The input wavfunction must have shape `(2,2,2,...)`. Attempt to factor the
+    wavefunction into a pure state over `keep_indices`. If this fails, a non-
+    unique mixture can be computed by eigendecomposition on the partial trace
+    of the input's density matrix.
+
+    States in the output mixture retain the shape (2,2,...).
+
+    Args:
+        wavefunction: A wavefunction to express over a qubit subset.
+        keep_indices: Which indices to express the wavefunction on.
+        atol: The tolerance for determining that a factored state is pure.
+
+    Returns:
+        A mixture representation of the partial trace of the wavefunction.
+    """
+
+    tot_dims = 1 << len(wavefunction.shape)
+    keep_dims = 1 << len(keep_indices)
+    try:
+        # Attempt to do efficient state factoring.
+        state = keep_qubits(wavefunction, keep_indices).reshape(keep_dims, 1)
+        return ((1.0, state))
+    except ValueError:
+        # Fall back to a (non-unique) mixture representation.
+        rho = np.kron(np.conj(wavefunction.reshape(tot_dims, 1)).T,
+                      wavefunction.reshape(tot_dims, 1)).reshape((2, 2) * len(wavefunction.shape))
+        keep_rho = partial_trace(rho, keep_indices).reshape((keep_dims,) * 2)
+        w, v = np.linalg.eig(keep_rho)
+        mixture = tuple(zip(w, [vec.reshape((2,) * len(keep_indices)) for vec in v.T]))
```

Comment:
This is the right zip ordering? It looks like you're zipping for each item instead of zipping the weights with the vectors.

Reply:
I've relabeled the variables for clarity